core/retries-api/src/main/java/software/amazon/awssdk/retriesapi/RetryStrategy.java

```diff
@@ -0,0 +1,208 @@
+        default Builder retryOnException(Class<? extends Throwable> throwable) {
+            return retryOnException(t -> t.getClass() == throwable);
+        }
+
+        /**
+         * Configure the strategy to retry when a failure exception class is an instance of the provided class (includes
+         * subtypes).
+         */
+        default Builder retryOnExceptionInstanceOf(Class<? extends Throwable> throwable) {
+            return retryOnException(t -> throwable.isAssignableFrom(t.getClass()));
+        }
+
+        /**
+         * Configure the strategy to retry when a failure exception or one of its cause classes is equal to the provided class.
+         */
+        default Builder retryOnExceptionOrCause(Class<? extends Throwable> throwable) {
+            return retryOnException(t -> {
+                if (t.getClass() == throwable) {
+                    return true;
+                }
+                Throwable cause = t.getCause();
+                while (cause != null) {
+                    if (cause.getClass() == throwable) {
+                        return true;
+                    }
+                    cause = cause.getCause();
+                }
+                return false;
+            });
+        }
+
+        /**
+         * Configure the strategy to retry when a failure exception or one of its cause classes is an instance of the provided
+         * class (includes subtypes).
+         */
+        default Builder retryOnExceptionOrCauseInstanceOf(Class<? extends Throwable> throwable) {
+            return retryOnException(t -> {
+                if (throwable.isAssignableFrom(t.getClass())) {
+                    return true;
+                }
+                Throwable cause = t.getCause();
+                while (cause != null) {
+                    if (throwable.isAssignableFrom(cause.getClass())) {
+                        return true;
+                    }
+                    cause = cause.getCause();
+                }
+                return false;
+            });
+        }
+
+        /**
+         * Configure the strategy to retry the root cause of a failure (the final cause) a failure exception is equal to the
+         * provided class.
+         */
+        default Builder retryOnRootCause(Class<? extends Throwable> throwable) {
+            return retryOnException(t -> {
+                boolean shouldRetry = false;
+                Throwable cause = t.getCause();
+                while (cause != null) {
+                    shouldRetry = throwable == cause.getClass();
+                    cause = cause.getCause();
+                }
+                return shouldRetry;
+            });
+        }
+
+        /**
+         * Configure the strategy to retry the root cause of a failure (the final cause) a failure exception is an instance of to
+         * the provided class (includes subtypes).
+         */
+        default Builder retryOnRootCauseInstanceOf(Class<? extends Throwable> throwable) {
```

Comment:
I guess that some of these conditions are subtle and error prone, having these spell out  with proper names make the API easier to use and understand, but I can remove some of these. Which ones do you think it will be worth keeping and which ones do you think we can go without?